### PR TITLE
Refactor Multinomial to support automatic inference of n_experiments, add UnnormalizedMultinomial (BagofCategoricals), simplify topic models

### PR DESCRIPTION
**Pull request by csy530216**
- Added multinomial option is_normalized to fit the case in LDA(not truly multinomial, but a production(multiplication) of variable number of categorical distribution. Also wrote some doc.
- Simplify lntm_mcem

However, the speed seems to be slower(e.g. 2.2s-->3.8s).

### CODE REVIEW

**Comment by thjashin** [comment]
Why there is nothing about "nosample" or "prob_only" anymore?

**Comment by thjashin** [comment]
@csy530216 Any update?

File: zhusuan/distributions/multivariate.py [hunk 1]
```diff
@@ -148,6 +157,102 @@ def _prob(self, given):
         return tf.exp(self._log_prob(given))
 
 
+class UnnormalizedMultinomial(Distribution):
+    """
+    The class of UnnormalizedMultinomial distribution.
+    Unnormalized multinomial distribution calculates probability in a different
+    way from multinomial distribution: It considers the bag-of-words `given` as
+    a statistics of a batch of ordered result sequences instead of a batch of
+    result counts. Hence it does not multiply the n!/(k1!*k2!*...) term.
```
Comment (thjashin) [inline zhusuan/distributions/multivariate.py]: n!/(k1!*k2!*...) should be made a math formula

File: zhusuan/distributions/multivariate.py [hunk 2]
```diff
@@ -148,6 +157,102 @@ def _prob(self, given):
         return tf.exp(self._log_prob(given))
 
 
+class UnnormalizedMultinomial(Distribution):
```
Comment (thjashin) [inline zhusuan/distributions/multivariate.py]: Maybe we could add an alias like `BagofCategorical`? How about this name? @cjf00000

File: zhusuan/distributions/multivariate.py [hunk 3]
```diff
@@ -148,6 +157,102 @@ def _prob(self, given):
...
+    """
+    The class of UnnormalizedMultinomial distribution.
+    Unnormalized multinomial distribution calculates probability in a different
```
Comment (thjashin) [inline zhusuan/distributions/multivariate.py]: probabilities

File: zhusuan/distributions/multivariate.py [hunk 4]
```diff
@@ -148,6 +157,102 @@ def _prob(self, given):
...
+    way from multinomial distribution: It considers the bag-of-words `given` as
```
Comment (thjashin) [inline zhusuan/distributions/multivariate.py]: :class:\`Multinomial\`

File: zhusuan/distributions/multivariate.py [hunk 5]
```diff
@@ -148,6 +157,102 @@ def _prob(self, given):
...
+
+    def _get_batch_shape(self):
+        if self.logits.get_shape():
+            return self.logits.get_shape()[:-1]
+        return tf.TensorShape(None)
+
+    def _sample(self, n_samples):
+        msg = "Unnormalized multinomial distribution does not support"
+        msg+= "sampling because n_experiments is not given. Please use"
+        msg+= "Multinomial to sample."
```
Comment (thjashin) [inline zhusuan/distributions/multivariate.py]: Just move the msg into Error then you can wrap easily.

File: examples/topic_models/lntm_mcem.py [hunk 6]
```diff
@@ -212,30 +201,30 @@ def _log_prior(observed):
...
                                           eta_mean: Eta_mean,
-                                          eta_logstd: Eta_logstd,
-                                          D_ph: _D,
-                                          n_chains_ph: _n_chains})
+                                          eta_logstd: Eta_logstd})
 
         time_ais += time.time()
 
         print('>> Test perplexity (AIS) ({:.1f}s)\n'
               '>> loglikelihood lower bound = {}\n'
```
Comment (thjashin) [inline examples/topic_models/lntm_mcem.py]: log likelihood

File: examples/topic_models/lntm_mcem.py [hunk 7]
```diff
@@ -212,30 +201,30 @@ def _log_prior(observed):
...
         ll_lb = _ais.run(sess, feed_dict={_x: X_test,
                                           eta_mean: Eta_mean,
-                                          eta_logstd: Eta_logstd,
-                                          D_ph: _D,
-                                          n_chains_ph: _n_chains})
+                                          eta_logstd: Eta_logstd})
 
         time_ais += time.time()
 
         print('>> Test perplexity (AIS) ({:.1f}s)\n'
```
Comment (thjashin) [inline examples/topic_models/lntm_mcem.py]: Test (AIS)

File: examples/topic_models/lntm_mcem.py [hunk 8]
```diff
@@ -139,16 +127,21 @@ def e_obj(observed):
...
     def _log_prior(observed):
-        log_p_eta, _, _ = joint_obj(observed)
-        return log_p_eta
+        return lntm(observed, _n_chains, _D, K, V,
+                    eta_mean, eta_logstd).local_log_prob('eta')
+
+    def _e_obj(observed):
+        model = lntm(observed, _n_chains, _D, K, V,
+                     eta_mean, eta_logstd)
+        return model.local_log_prob('eta') + model.local_log_prob('x')
```
Comment (thjashin) [inline examples/topic_models/lntm_mcem.py]: Can we reuse the above `e_obj` since the only difference is in `D` and `n_chains`.

File: examples/topic_models/lntm_mcem.py [hunk 9]
```diff
@@ -15,33 +15,37 @@
...
-                            tf.tile(tf.expand_dims(eta_mean, 0), [D, 1]),
-                            0), [n_chains, 1, 1]),
-                        logstd=eta_logstd,
-                        group_ndims=1)
-        beta = zs.Normal('beta', tf.zeros([K, V]), logstd=log_delta,
-                         group_ndims=1)
-        theta = tf.reshape(tf.nn.softmax(eta), [-1, K])
+        eta_mean = tf.tile(tf.expand_dims(
+                           tf.tile(tf.expand_dims(eta_mean, 0), [n_docs, 1]),
+                           0), [n_chains, 1, 1])
```
Comment (thjashin) [inline examples/topic_models/lntm_mcem.py]: Can we do
```python
eta_mean = tf.tile(tf.expand_dims(eta_mean, 0), [n_docs, 1])
eta = zs.Normal('eta', eta_mean, logstd=eta_logstd, group_ndims=1, n_samples=n_chains)
```

File: examples/topic_models/lntm_mcem.py [hunk 10]
```diff
@@ -15,33 +15,37 @@
...
+        eta = zs.Normal('eta', eta_mean, logstd=eta_logstd, group_ndims=1)
+        theta = tf.nn.softmax(eta)
+        # beta/phi: Unnormalized/normalized topic-word matrix
+        beta = zs.Normal('beta', tf.zeros([n_topics, n_vocab]),
+                         logstd=log_delta, group_ndims=1)
         phi = tf.nn.softmax(beta)
-        pred = tf.matmul(theta, phi)
-        pred = tf.reshape(pred, tf.stack([n_chains, D, V]))
-        x = zs.UnnormalizedMultinomial('x', tf.log(pred),
+        # doc_word: (Automatically) normalized document-word matrix
```
Comment (thjashin) [inline examples/topic_models/lntm_mcem.py]: remove "(Automatically) normalized"

File: zhusuan/distributions/utils.py [hunk 11]
```diff
@@ -282,22 +277,21 @@ def assert_positive_int32(value, name, need_scalar):
             raise TypeError(name + " must be integer")
         elif value <= 0:
             raise ValueError(name + " must be positive")
+    elif isinstance(value, (list, tuple, dict)):
+        raise TypeError(name + " should be a scalar (0-D Tensor).")
```
Comment (thjashin) [inline zhusuan/distributions/utils.py]: remove this check, convert_to_tensor and assert_rank will raise in that case.

Comment (csy530216) [reply zhusuan/distributions/utils.py]: I write the check because the doc says
> If `value` is the instance of built-in type, it will be checked directly.
Otherwise, it will be converted to a `tf.int32` tensor and checked.

Comment (thjashin) [reply zhusuan/distributions/utils.py]: I realize it's not very easy to check if a type is a numeric type in python. And there is actually no need that we keep n_experiments a int. So it would be better that we leave all things to TF. We should remove all things in the if and do the convert directly.

**Comment by thjashin** [comment]
LGTM. Any to change before merging? @csy530216

**Comment by csy530216** [comment]
I think that's okay.
